browser/ui/views/toolbar/bookmark_button.cc

```diff
@@ -0,0 +1,98 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/browser/ui/views/toolbar/bookmark_button.h"
+
+#include "base/strings/utf_string_conversions.h"
+#include "chrome/app/chrome_command_ids.h"
+#include "chrome/browser/ui/view_ids.h"
+#include "chrome/browser/themes/theme_properties.h"
+#include "chrome/grit/generated_resources.h"
+#include "components/strings/grit/components_strings.h"
+#include "components/toolbar/vector_icons.h"
+#include "ui/accessibility/ax_node_data.h"
+#include "ui/base/l10n/l10n_util.h"
+#include "ui/base/theme_provider.h"
+#include "ui/gfx/paint_vector_icon.h"
+#include "ui/views/widget/widget.h"
+
+BookmarkButton::BookmarkButton(views::ButtonListener* listener)
+    : ToolbarButton(listener),
+      widget_observer_(this) {
+        set_id(VIEW_ID_STAR_BUTTON);
+        set_tag(IDC_BOOKMARK_PAGE);
+        SetAccessibleName(l10n_util::GetStringUTF16(IDS_ACCNAME_FORWARD));
+      }
+
+BookmarkButton::~BookmarkButton() {
+}
+
+const char* BookmarkButton::GetClassName() const {
+  return "BookmarkButton";
+}
+
+bool BookmarkButton::GetTooltipText(const gfx::Point& p,
+                                  base::string16* tooltip) const {
+  int textId = active_ ? IDS_TOOLTIP_STARRED
+                                      : IDS_TOOLTIP_STAR;
+  tooltip->assign(l10n_util::GetStringUTF16(textId));
+  return true;
+}
+
+void BookmarkButton::GetAccessibleNodeData(ui::AXNodeData* node_data) {
+  int textId = active_ ? IDS_TOOLTIP_STARRED
+                                      : IDS_TOOLTIP_STAR;
+    node_data->role = ax::mojom::Role::kButton;
+  node_data->SetName(l10n_util::GetStringUTF16(textId));
+}
+
+void BookmarkButton::SetHighlighted(bool bubble_visible) {
+  AnimateInkDrop(bubble_visible ? views::InkDropState::ACTIVATED
+                                : views::InkDropState::DEACTIVATED,
+                 nullptr);
+}
+
+void BookmarkButton::SetToggled(bool on) {
+
+  active_ = on;
+
+  const ui::ThemeProvider* tp = GetThemeProvider();
+  const ui::NativeTheme* theme = GetNativeTheme();
+
+  SkColor icon_color =
+      on ? theme->GetSystemColor(
+                    ui::NativeTheme::kColorId_ProminentButtonColor)
+          : tp->GetColor(ThemeProperties::COLOR_TOOLBAR_BUTTON_ICON);
+  SetImage(views::Button::STATE_NORMAL, gfx::CreateVectorIcon(
+      toolbar::kStarIcon, icon_color));
```

Comment:
Maybe we can just use our own vector icon here instead of the patch below to use his one?

Reply:
I'm a little confused about this whole class because the description sounds like we are just moving it, so why do we need a new class? Can't we just re-use the existing StarView in a different place?

Reply:
@bbondy we are not patching the icon - it's our own icon file, I just needed to add a different size variant to the file.

@bridiver `StarView` inherits `PageActionIconView`, which expects to be inside the `LocationBarView` visually and hierarchically. My first attempt at this did re-use that class, but it came with some side effects (mainly visual). The visual aim is to have this be another top-level toolbar button, which `ToolbarButton` achieves without any extra rendering code. The modifications to `PageActionIconView` may have been numerable and would be repeating visual style, making it potentially more difficult to rebase in the future than using the simple `ToolbarButton` class.

Reply:
How about adding new `UpdateIconImage()` with codes from line 60 to last and calling it when theme changing event occurs because this button's color depends on theme property.